apps/backend/src/libs/modules/open-ai/open-ai.module.ts

```diff
@@ -0,0 +1,276 @@
+			const assistant = existingAssistants.data.find(
+				(assistant) => assistant.name === OpenAIAssistantConfig.NAME,
+			);
+
+			const initializedAssistant =
+				assistant ??
+				(await this.openAi.beta.assistants.create({
+					instructions: OpenAIAssistantConfig.INSTRUCTION,
+					model: this.config.ENV.OPEN_AI.MODEL,
+					name: OpenAIAssistantConfig.NAME,
+					temperature: OpenAIAssistantConfig.TEMPERATURE,
+					tools: [...OpenAIAssistantConfig.TOOLS],
+					top_p: OpenAIAssistantConfig.TOP_P,
+				}));
+
+			this.logger.info(
+				`AI Assistant "${initializedAssistant.name as string}" is fully ready for interaction.`,
+			);
+			this.logger.info(`AI Assistant model - "${initializedAssistant.model}".`);
+
+			return initializedAssistant.id;
+		} catch (error) {
+			this.logger.error(`Error initializing AI assistant: ${String(error)}`);
+
+			throw new OpenAIError({
+				message: OpenAIErrorMessage.WRONG_RESPONSE,
+				status: HTTPCode.INTERNAL_SERVER_ERROR,
+			});
+		}
+	}
+
+	private async handleRequiresAction(
+		run: OpenAI.Beta.Threads.Runs.Run,
+		functionName: string,
+	): Promise<void> {
+		if (!run.required_action) {
+			return;
+		}
+
+		const { tool_calls } = run.required_action.submit_tool_outputs;
+
+		await Promise.all(
+			tool_calls.map(
+				async ({
+					function: toolFunction,
+					function: { arguments: toolArguments },
+					id,
+				}) => {
+					if (toolFunction.name !== functionName) {
+						return;
+					}
+
+					try {
+						await this.openAi.beta.threads.runs.submitToolOutputsAndPoll(
+							run.thread_id,
+							run.id,
+							{
+								tool_outputs: [
+									{
+										output: toolArguments,
+										tool_call_id: id,
+									},
+								],
+							},
+						);
+					} catch (error) {
+						this.logger.error(
+							`Error submitting tool outputs: ${String(error)}`,
+						);
+
+						throw new OpenAIError({
+							message: OpenAIErrorMessage.WRONG_RESPONSE,
+							status: HTTPCode.INTERNAL_SERVER_ERROR,
+						});
+					}
+				},
+			),
+		);
+	}
+
+	private async handleRunStatus(
+		threadId: string,
+		run: OpenAI.Beta.Threads.Runs.Run,
+		functionName: string,
+	): Promise<OpenAiResponseMessage> {
+		if (run.status === OpenAIRunStatus.COMPLETED) {
+			return await this.getAllMessages(threadId);
+		} else if (run.status === OpenAIRunStatus.REQUIRE_ACTIONS) {
+			await this.handleRequiresAction(run, functionName);
+
+			return await this.getAllMessages(threadId);
+		} else {
+			this.logger.error(`AI Assistant run failed: ${run.status}`);
+
+			throw new OpenAIError({
+				message: OpenAIErrorMessage.WRONG_RESPONSE,
+				status: HTTPCode.INTERNAL_SERVER_ERROR,
+			});
+		}
+	}
+
+	public async addMessageToThread(
+		threadId: string,
+		message: OpenAiRequestMessage,
+	): Promise<boolean> {
+		try {
+			const newMessage = await this.openAi.beta.threads.messages.create(
+				threadId,
+				message,
+			);
+
+			return !("error" in newMessage);
```

Comment:
As you have a couple of places with ```"error" in object``` check. Let's move it to a type guard isError to make the intention of the code clearer (here you want to check if the object of a message is an error)

Reply:
I tried to implement Type Gyard but OpenAI returns an "error" response with type which is hard to handle in Type Gyard.